List os object names (#136)
PowerShell/Community-Blog: Posts/2023/09/List-OS-Object-Names.md

---

As our knowledge of the Windows operating system grows, we often need to understand how processes
request, utilize, and dispose of operating system resources. This is nothing new, tools like
the Task Manager or Process Explorer shows a comprehensive view of these relations.
However, we don't need a complicated application to explore this lower level of abstraction, all
we need is our trusty PowerShell console, and a bit of imagination.

## Method

To achieve this, and to keep it fun, we are going to use a set of documented and partially
documented Windows APIs, and although it's possible to achieve this using 100% PowerShell, to
improve simplicity we are going to use a bit of C#, just for the function signatures, and
value types, like structs.
This post draws inspiration from a blog post by Pavel Yosifovich, that can me accessed [here][01].
This is one of the posts I keep in my browser's bookmarks.

## Signatures

As explained previously, we are going to define function signatures, and structs using C#. Due to
similarities between C# and PowerShell, this shouldn't be complicated to understand.

> ```csharp

Author: I've split the code and added comments to make it more understandable. Thanks for the suggestion.

---

As our knowledge of the Windows operating system grows, we often need to understand how processes
request, utilize, and dispose of operating system resources. This is nothing new, tools like
the Task Manager or Process Explorer show a comprehensive view of these relations.
However, we don't need a complicated application to explore this lower level of abstraction. All
we need is our trusty PowerShell console, and a bit of imagination.

## Method

To achieve this, and to keep it fun, we are going to use a set of documented and partially
documented Windows APIs. Although it's possible to achieve this using 100% PowerShell, to improve
simplicity we are going to use a bit of C# for the function signatures and value types, like
structs.
This post draws inspiration from the blog post [How can I close a handle in another process?][01]
by Pavel Yosifovich. This is one of the posts I keep in my browser's bookmarks.

## Signatures

As explained previously, we are going to define function signatures, and structs using C#. Due to
similarities between C# and PowerShell, this should be relatively familiar.

> First we start with the enumerations. These are for making function calls more understandable.

Reviewer: ```suggestion
First we start with the enumerations. These are for making function calls more understandable.
Define a new file named `Utilities.Enums.cs` for the enum definitions.
```

```csharp

> namespace Utilities

Reviewer: ```suggestion
// Utilities.Enums.cs
namespace Utilities
```

{
    using System;
    using System.Runtime.InteropServices;

    // This one lists the possible process information we can get
    // with 'NtQueryInformationProcess'.
    public enum PROCESSINFOCLASS
    {
        ProcessBasicInformation = 0,
        ProcessDebugPort = 7,
        ProcessWow64Information = 26,
        ProcessImageFileName = 27,
        ProcessBreakOnTermination = 29,
        ProcessHandleInformation = 51
    }

    // This one is similar to the previous enumeration, but lists
    // possible information we can get from objects.
    public enum OBJECT_INFORMATION_CLASS
    {
        ObjectNameInformation = 1,
        ObjectTypeInformation = 2

>     }
> 
>     /* Code continues bellow */
> }

Reviewer: ```suggestion
    }
}
```

```

> Now we define the structs. There are the structures that will hold the information
> we want to query.

Reviewer: ```suggestion
Now we define the structs. There are the structures that will hold the information
we want to query. Create a new file named `Utilities.Structs.cs`.
```

```csharp

> namespace Utilities

Reviewer: ```suggestion
// Utilities.Structs.cs
namespace Utilities
```

{
    using System;

>     using System.Runtime.InteropServices;
> 
>     /* Above code */
> 
>     // Often when using functions from 'ntdll.dll' we see this

Reviewer: ```suggestion
    using System.Runtime.InteropServices;

    // Often when using functions from 'ntdll.dll' we see this
```

        public GENERIC_MAPPING GenericMapping;
        public uint ValidAccessMask;
        public bool SecurityRequired;
        public bool MaintainHandleCount;
        public byte TypeIndex;
        public char ReservedByte;
        public uint PoolType;
        public uint DefaultPagedPoolCharge;
        public uint DefaultNonPagedPoolCharge;
    }

    // This structure holds information about a process open
    // handles. An array of these structures is what we are
    // going to work with.
    [StructLayout(LayoutKind.Sequential)]
    public struct PROCESS_HANDLE_TABLE_ENTRY_INFO
    {
        public IntPtr HandleValue;
        public ulong HandleCount;
        public ulong PointerCount;
        public uint GrantedAccess;
        public uint ObjectTypeIndex;
        public uint HandleAttributes;
        public uint Reserved;
    }

    // This structure is what is returned when we use the option
    // 'ProcessHandleInformation' when querying 'NtQueryInformationProcess'
    [StructLayout(LayoutKind.Sequential)]
    public struct PROCESS_HANDLE_SNAPSHOT_INFORMATION
    {
        public ulong NumberOfHandles;
        public ulong Reserved;
        public IntPtr Handles;

>     }
> 
>     /* Code continues bellow */
> 
> }

Reviewer: ```suggestion
    }
}
```

```

> Lastly we define the functions.

Reviewer: ```suggestion
Lastly, we define the functions in a file named `Utilities.Members.cs`.
```

```csharp

> namespace Utilities

Reviewer: ```suggestion
// Utilities.Members.cs
namespace Utilities
```

{
    using System;

>     using System.Runtime.InteropServices;
> 
>     /* Above code */
> 
>     // We need a class to hold members like methods (functions).

Reviewer: ```suggestion
    using System.Runtime.InteropServices;

    // We need a class to hold members like methods (functions).
```

        // This function duplicates a handle opened by a process.
        // We need to duplicate the handle to be able to query it.
        [DllImport("kernel32.dll", SetLastError=true)]
        [return: MarshalAs(UnmanagedType.Bool)]
        public static extern bool DuplicateHandle(
            IntPtr hSourceProcessHandle,
            IntPtr hSourceHandle,
            IntPtr hTargetProcessHandle,
            out IntPtr lpTargetHandle,
            uint dwDesiredAccess,
            [MarshalAs(UnmanagedType.Bool)] bool bInheritHandle,
            uint dwOptions
        );

        // This function returns a handle to the current process.
        // We need it so the OS can associate the duplicated handle
        // with it.
        [DllImport("kernel32.dll", SetLastError = true)]
        public static extern IntPtr GetCurrentProcess();

        // This function closes a handle opened by certain functions.
        // Is important to close all handles where necessary to avoid
        // memory leaks.
        [DllImport("kernel32.dll", SetLastError = true)]
        public static extern bool CloseHandle(IntPtr hObject);
    }
}
```

## The script

### Adding the signature type

> The first thing we need to do is add that signature as a type in our PowerShell script. There are
> two ways we can do it, by copying that C# text into a here-string, or saving it in a file.
> The second one keeps the code cleaner, so for this post, let's assume I saved that file as
> `C:\HandleObjectNames.cs`.

Reviewer: ```suggestion
The first thing we need to do is add that signature as a type in our PowerShell script. Because we
defined the code in separate files, we can use the `Add-Type` cmdlet with the **Path** parameter to
compile them together and add the `Utilities` namespace.
```

```powershell
try {

>     Add-Type -TypeDefinition (Get-Content -Path 'C:\HandleObjectNames.cs' -Raw)

Reviewer: ```suggestion
    Add-Type -Path @(
        './Utilities.Enums.cs'
        './Utilities.Structs.cs'
        './Utilities.Members.cs'
    )
```

            $obj_type_buffer,             # The buffer.
            1024,                         # The buffer size.
            [ref]$bytes_needed            # The size needed.
        )
        
        # We always need to close our handle, regardless of success or failure.
        [void][Utilities.ProcessAndThread]::CloseHandle($h_target)

        # Something went wrong, continuing to the next handle.
        if ($result -ne -0) { continue }

        # Marshaling the buffer, and adding the object name to our list.
        $obj_type_info = [System.Runtime.InteropServices.Marshal]::PtrToStructure(
            $obj_type_buffer, [type][Utilities.OBJECT_TYPE_INFORMATION])

        if (!$object_names.Contains($obj_type_info.TypeName.Buffer)) {
            [void]$object_names.Add($obj_type_info.TypeName.Buffer)
        }

        # Increasing the offset to the next handle.
        $offset = [IntPtr]::Add($offset, [IntPtr]::Size)
    }
}

# Returning the result.
return $object_names
```

And this is how we leverage the operating system's APIs to query object names.
This goes to show how powerful PowerShell really is. I hope you had fun, and until the next one!

Francisco Nabas.

> [OpenProcess function][02]  
> [DuplicateHandle function][03]  
> [NtQueryInformationProcess function][04]  
> [NtQueryObject function][05]
> 
> Check out my PowerShell modules!  
> Manage Windows like a pro with [WindowsUtils][06]  
> Check with module/assembly failed to load with [LibSnitcher][07]

Reviewer 2: ```suggestion

- [OpenProcess function][02]  
- [DuplicateHandle function][03]  
- [NtQueryInformationProcess function][04]  
- [NtQueryObject function][05]

Check out my PowerShell modules!  

- Manage Windows like a pro with [WindowsUtils][06]  
- Check with module/assembly failed to load with [LibSnitcher][07]

```
